src/main/java/com/google/codeu/data/Datastore.java

```diff
@@ -80,6 +81,24 @@ public void storeMessage(Message message) {
 
     return convertEntitiesToMessages(results);
   }
+  
+  /**
+   * Gets messages posted by a specific user.
+   *
+   * @return a list of messages posted by user, or empty list if no messages posted to country page.
+   *     List is sorted by time descending.
+   */
+  public List<Message> getMessagesByCategory(String category){
+	Query query =
+	    new Query("Message")
+	        .setFilter(new Query.FilterPredicate("category", FilterOperator.EQUAL, category))
```

Comment:
Maybe I'm misunderstanding, but my instinct is that we want to  "getCountryMessagesByCategory" to search for messages on one country in one category? So that on the country's category page, we are only showing that particular country's category?

Reply:
It seems to be working correctly, so maybe I just don't understand how it's working?

Reply:
Yes, that was the intention! Although I don't think I ended up using that function. And without a country parameter, I don't think it would work properly. I'll go ahead and take it out.

Reply:
So there should be a country parameter in the message entity, and we can just add that to the filter. How is it working now? I'm not sure that I understand how you are querying the messages for the category page right now

Reply:
So there's a messages attribute that was already set through the query in the CountryServlet. That attribute is forwarded to the CategoryServlet (nothing is done with it), and then to the category.jsp. I also stored a category attribute. In the jsp, I use a loop to iterate through the messages, and only print the ones that have their category matching the category attribute. It might be less efficient than just doing a query based on the individual category though, but messages was already an attribute, so I figured I'd just use that.

Reply:
I just noticed some debugging in the jsp, I'll take that out also!

Reply:
Ohhhhh, I see. That make sense. 

If you can call the category servlet before the messages are loaded in countryservlet, though, it might be more efficient to do query with multiple parameters, as I think it would be O(n) vs O(nm), where m is country messages. However, if it's not possible that's okay. 
I think, though, the syntax should be like 
```
Filter lastNameFilter = new FilterPredicate("lastName", FilterOperator.EQUAL, targetLastName);

Filter cityFilter = new FilterPredicate("city", FilterOperator.EQUAL, targetCity);

Filter birthYearMinFilter =
    new FilterPredicate("birthYear", FilterOperator.GREATER_THAN_OR_EQUAL, minBirthYear);

Filter birthYearMaxFilter =
    new FilterPredicate("birthYear", FilterOperator.LESS_THAN_OR_EQUAL, maxBirthYear);

Filter validFilter =
    CompositeFilterOperator.and(
        lastNameFilter, cityFilter, birthYearMinFilter, birthYearMaxFilter);

Query q = new Query("Person").setFilter(validFilter);
```
according to [the query docs](https://cloud.google.com/appengine/docs/standard/java/datastore/query-restrictions)
which I think would be something like 
```
Query.Filter countryFilter = new Query.FilterPredicate("country", FilterOperator.EQUAL, countryCode);
Query.Filter categoryFilter = new Query.FilterPredicate("category", FilterOperator.EQUAL, category);

Query.Filter combinedFilter = Query.CompositeFilterOperator.and(countryFilter, categoryFilter);

Query query = 
       new Query("Message")
            .setFilter(combinedFilter);
            .addSort("timestamp", SortDirection.DESCENDING);
    PreparedQuery results = datastore.prepare(query);
    Entity userEntity = results.asSingleEntity();
```

*Full disclosure, I only know to look for this because deleting messages was really just one big fight with the query documentation*

Reply:
LOL You're a life saver for having me avoid a big fight with query documentation, and for explaining everything, THANK YOU!!

Reply:
Am I using userEntity for anything?

Reply:
Oh whoops, I forgot to rename that. I think it would be the messageEntity like how we currently search for messages, and then you convert that messageEntity to a list of messages and return those. I think.

Reply:
I think it works without adding the results.asSingleEntity(); What exactly does that line do? I fed in results to convertEntitiestoMessages and returned that, like is done in getMessagesByUser. Do you think that's okay?

Reply:
I think it probably is! Chances are, asSingleEntity() was for returning only one message